### ISSUE TITLE
No cgroup filesystem mounted on /sys/fs/cgroup

### ISSUE THREAD

**Issue by marwinski** [gardenlinux/gardenlinux#275]
**What happened**:

We have several instances where the `/sys/fs/cgroup` mount mysteriously disappeared. This does not only happen on Garden Linux but with other linux distributions as well (with slightly older systemd and kernel).

**What you expected to happen**:

I don't expect this to happen.

**How to reproduce it (as minimally and precisely as possible)**:

We have seen roughly 3 instances of this last week. We don't know how to reproduce it or what causes it.

**Anything else we need to know**:

Garden Linux 184.0 on AWS and GCP.

These are some symptoms that we saw so far:

```
$ systemd-cgls
Failed to get root control group path.
No cgroup filesystem mounted on /sys/fs/cgroup
Failed to list cgroup tree: No medium found
```

The nodes appear healthy but all hell breaks loose on the Kubernetes notes, see output 

```
$ k -n kube-system get po calico-node-pc7l9 -o yaml

status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:11:23Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:10:37Z"
    message: 'containers with unready status: [calico-node]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:10:37Z"
    message: 'containers with unready status: [calico-node]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2021-06-21T06:10:37Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://392b39106489f59c36ee7a71da89dffaf744a76dc09e7d7fa79ccc66b88618b4
    image: eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/node@sha256:b0997b0a69657aa70da903c3f6c88adb6bd517a78c2adc3e7174a3f94f212375
    imageID: docker-pullable://eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/node@sha256:b0997b0a69657aa70da903c3f6c88adb6bd517a78c2adc3e7174a3f94f212375
    lastState:
      terminated:
        containerID: docker://392b39106489f59c36ee7a71da89dffaf744a76dc09e7d7fa79ccc66b88618b4
        exitCode: 128
        finishedAt: "2021-06-21T06:11:29Z"
        message: 'OCI runtime create failed: container_linux.go:370: starting container
          process caused: process_linux.go:459: container init caused: rootfs_linux.go:117:
          jailing process inside rootfs caused: pivot_root invalid argument: unknown'
        reason: ContainerCannotRun
        startedAt: "2021-06-21T06:11:29Z"
    name: calico-node
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        message: 'failed to start container "392b39106489f59c36ee7a71da89dffaf744a76dc09e7d7fa79ccc66b88618b4":
          Error response from daemon: OCI runtime create failed: container_linux.go:370:
          starting container process caused: process_linux.go:459: container init
          caused: rootfs_linux.go:117: jailing process inside rootfs caused: pivot_root
          invalid argument: unknown'
        reason: RunContainerError
  hostIP: 10.250.21.205
  initContainerStatuses:
  - containerID: docker://5168fdf5831aa9b299a95c3fdc0a0bc91a2374b56fc4ffd956607c172f86b665
    image: eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/cni@sha256:4ca9a473d78e4d06ca7871c7c926cc8a8dd9a181fd87aee89a791db831045f8f
    imageID: docker-pullable://eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/cni@sha256:4ca9a473d78e4d06ca7871c7c926cc8a8dd9a181fd87aee89a791db831045f8f
    lastState: {}
    name: install-cni
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: docker://5168fdf5831aa9b299a95c3fdc0a0bc91a2374b56fc4ffd956607c172f86b665
        exitCode: 0
        finishedAt: "2021-06-21T06:10:49Z"
        reason: Completed
        startedAt: "2021-06-21T06:10:47Z"
  - containerID: docker://4d906f875fdefa08127c98d9676b2dc9eec432a66d07b9aece1d84c9f77ab1b3
    image: eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/pod2daemon-flexvol@sha256:1160693a7dfbb8c4512faff9007cc879a65824dc0f05fbb2bc4acc1b4ba225a4
    imageID: docker-pullable://eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/pod2daemon-flexvol@sha256:1160693a7dfbb8c4512faff9007cc879a65824dc0f05fbb2bc4acc1b4ba225a4
    lastState: {}
    name: flexvol-driver
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: docker://4d906f875fdefa08127c98d9676b2dc9eec432a66d07b9aece1d84c9f77ab1b3
        exitCode: 0
        finishedAt: "2021-06-21T06:11:10Z"
        reason: Completed
        startedAt: "2021-06-21T06:11:10Z"
  - containerID: docker://c9841568f73149a8ce90a99a1c66cbb658f05ac7579251a730f05294d44dd4d1
    image: sha256:dfd78094c8f5fcf8b698da3e9d1032cc7e76f2c28642802b48c221d676f32625
    imageID: docker-pullable://eu.gcr.io/sap-se-gcr-k8s-public/docker_io/calico/pod2daemon-flexvol@sha256:1160693a7dfbb8c4512faff9007cc879a65824dc0f05fbb2bc4acc1b4ba225a4
    lastState: {}
    name: install-cni-check
    ready: true
    restartCount: 0
    state:
      terminated:
        containerID: docker://c9841568f73149a8ce90a99a1c66cbb658f05ac7579251a730f05294d44dd4d1
        exitCode: 0
        finishedAt: "2021-06-21T06:11:16Z"
        reason: Completed
        startedAt: "2021-06-21T06:11:16Z"
  phase: Running
  podIP: 10.250.21.205
  podIPs:
  - ip: 10.250.21.205
  qosClass: Burstable
  startTime: "2021-06-21T06:10:37Z"
```

```
$ cat /etc/os-release
PRETTY_NAME="Garden Linux 184.0"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://gardenlinux.io/"
SUPPORT_URL="https://github.com/gardenlinux/gardenlinux"
BUG_REPORT_URL="https://github.com/gardenlinux/gardenlinux/issues"
GARDENLINUX_FEATURES=base,aws,_nopkg,_prod,_readonly,_slim,base,cloud,gardener,server
GARDENLINUX_VERSION=184.0
GARDENLINUX_COMMIT_ID=e5a1576deded4cb778b178ed279c38138e384f63
VERSION_CODENAME=bullseye
```

On a healthy node (this is expected):

```
# cat /etc/mtab | grep cgroup
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,size=4096k,nr_inodes=1024,mode=755 0 0
cgroup2 /sys/fs/cgroup/unified cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
tmpfs /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup tmpfs rw,nosuid,nodev,noexec,relatime,mode=755 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,name=systemd 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/rdma cgroup rw,nosuid,nodev,noexec,relatime,rdma 0 0
cgroup /var/lib/docker/overlay2/fc4fae65544160cc1f9696e11d7d36adaac1ce6d1b51605454ff30f2591f5678/merged/sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
```

On the problematic Node:

```
# cat /etc/mtab | grep cgroup
#
```

We have looked at some logs and saw the following:

- The problem appeared shortly after the node was instantiated and started. We don't know whether it would happen after a reboot.
- It is our understanding that systemd mounts cgroupfs. We however don't see that in the logs but assume this from code analysis.
- There is no indication that the cgroupfs has been forcefully unmounted. In any case this would be difficult as it is quite "busy"
- Initially the cgroupsfs is mounted, kubelet reports it and docker does not complain

```
2021-06-24T10:27:35.140171+00:00 localhost kubelet[4059]: I0624 10:27:35.139795    4059 fs.go:126] Filesystem partitions: map[/dev/sda3:{mountpoint:/boot major:8 minor:3 fsType:xfs blockSize:0} /dev/sda4:{mountpoint:/ major:8 minor:4 fsType:xfs blockSize:0} /dev/shm:{mountpoint:/dev/shm major:0 minor:23 fsType:tmpfs blockSize:0} /run:{mountpoint:/run major:0 minor:25 fsType:tmpfs blockSize:0} /sys/fs/cgroup:{mountpoint:/sys/fs/cgroup major:0 minor:26 fsType:tmpfs blockSize:0} /var/lib/docker/containers/4ecf6fc8f8b7bff1d44df7151f04a9f39de32b6d91d2921802ff6160c255a499/mounts/secrets:{mountpoint:/var/lib/docker/containers/4ecf6fc8f8b7bff1d44df7151f04a9f39de32b6d91d2921802ff6160c255a499/mounts/secrets major:0 minor:46 fsType:tmpfs blockSize:0} overlay_0-45:{mountpoint:/var/lib/docker/overlay2/a56c52bc81fd9af461a004cd7f997accbaefc3010714d9a50292528572a0e9d0/merged major:0 minor:45 fsType:overlay blockSize:0}]
```

**Environment**:

**Comment by marwinski** [comment]
This is the result of the analysis done so far. We have not been able to pin down the root cause for this behavior but have some leads to work on.

So far we have seen this issue with various versions:
- Linux kernel 5.3 and 5.4
- Docker 19 and Docker 20
- Kubernetes 18 and 19

We have seen the issue roughly 4 times in 2 weeks. We were able to secure a disk snapshot of a broken node but unfortunately some important logs had been overwritten so that this yielded very little additional information. While this problem renders a node unusable it is reported healthy by the kubelet so manual intervention is required to recover it. In addition we also think that there are a number of unreported cases.

There is absolutely no evidence in any log that a component forcefully unmounted the `/sys/fs/cgroup` file system.

This is what we have considered and investigated so far:

## (1) Early restart of the kubelet

During node startup the kubelet is restarted while starting up for the first time. This is due to a bug between Gardener and the OS extensions (same issue for Garden Linux and SuSE) - systemd initialization is done twice.

While this is awkward, we see the possibility that this causes the issue as low (e.g. due to a race condition or incomplete cleanup). The double restart issue will be fixed.

## (2) Kubelet uses cgroupfs as cgroup driver

[This proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/kubelet-systemd.md#cgroupfs-single-writer) states that in systemd systems (like Garden Linux and SuSE) cgroup management should be done by systemd and not cgroupfs as we do it in Gardener. 

While this is awkward and may lead to subtle issues, we consider it unlikely to be the root cause of the issue.

## (3) Deployed content

We see that the issue shows up during initial startup of the node while system containers are being started. We looked at the containers and the calico-node container is the only container where the content would be able to unmount the cgroup filesystem due to the following configuration in the calico-node container (and only in this one):

```
  volumes:
  - hostPath:
      path: /sys/fs/
      type: DirectoryOrCreate
    name: sysfs
```

and 

```
    volumeMounts:
    - mountPath: /sys/fs/
      mountPropagation: Bidirectional
      name: sysfs
```

We validated that with the `bidirectional` mount the cgroup file system can be unmounted. This is not possible in a privileged container without that mount.

It is suspicious that in all cases that we were able to investigate the calico-node init containers completed successfully but the calico-node container did not. In one instance it started but failed after 2 seconds.

We have done some source code analysis and we did not find any evidence that some piece of code in calico is responsible for the unmount operation. Unfortunately we were not be able to collect logs that could either prove nor disprove this.

Gardener currently does not make use of the eBPF features of Calico which is the reason for the bidirectional mount. We have therefore made the recommendation to remove the mount and disable eBPF.

## (4) Kubelet 

The kubelet can trigger an unmount operation that would cause this issue. The Kubelet source code shows that unmount operations are logged with a log level that would show up in the system log. We did not find such log messages in our logs and therefore consider the kubelet to be an unlikely cause.

## (5) Docker Container Runtime

The container runtime has ample code that unmounts file systems. No investigation has been done so far as we consider the container runtime as an unlikely cause for the issue.

## (6) Linux Kernel Issue

Based on discussions with a Linux Kernel developer we consider it unlikely that this is a kernel issue. We have seen this issue with quite different kernels.

# Next Steps

All mitigations will take time and we will not be able to learn immediately whether they will resolve the issue. This said, we did not find a likely root cause for the isusue so far.

## 1. Unmount Tracing

We will investigate means to trace unmount operations and to add some functionality during early startup that should show the source of such unmount operation.

## 2. Follow up on Detected issues

We will follow up on the following:
1. Restart of the kubelet during the early boot process
2. Change from croupfs to systemd for cgroup management
3. Remvoe the bidirectional mount from Calico

**Comment by mbanck** [comment]
One thing one could do (as pointed out by @waldiTM) would be to use auditd to monitor/log mounts/unmounts, this might shed some light on this (but could also lead to non-negligible additional logging due to containers (bind-)mounting so many things these days), see https://www.thegeekdiary.com/how-to-monitor-the-mounting-umounting-of-mount-points-using-auditd-on-centos-rhel-67/

**Comment by marwinski** [comment]
Removal of the bidirectional mount in gardener calico network extension: https://github.com/gardener/gardener-extension-networking-calico/pull/102

**Comment by marwinski** [comment]
We have worked on the following:

If we are in a situation with a missing cgroup mount a `systemctl daemon-reexec` will remount the cgroup directory but not all the controllers. This does not work as a remedy. In one instance a reboot of a broken node was successful: the node came up and joined the cluster. We have drafted a script that will be added to he Garden Linux extension. This will 
- check once per second for a missing /var/fs/cgroup for the first ten minutes after startup
- in case of a missing cgroup directory it will save forensic information and reboot
 
We ran this manually and could identify the container that triggered the unmount operation based on the forensic information.
